Mathlib/RingTheory/Localization/Basic.lean

```diff
@@ -0,0 +1,1405 @@
+          refine' mk_eq_mk_iff.mpr (r_of_eq _)
+          simp only [Submonoid.coe_mul]
+          ring) }
+
+theorem sub_mk (a c) (b d) : (mk a b : Localization M) - mk c d =
+    mk ((d : R) * a - b * c) (b * d) :=
+  calc
+    mk a b - mk c d = mk a b + -mk c d := sub_eq_add_neg _ _
+    _ = mk a b + mk (-c) d := by rw [neg_mk]
+    _ = mk (b * -c + d * a) (b * d) := (add_mk _ _ _ _)
+    _ = mk (d * a - b * c) (b * d) := by congr; ring
+
+#align localization.sub_mk Localization.sub_mk
+
+theorem mk_int_cast (m : ℤ) : (mk m 1 : Localization M) = m := by
+  simpa using @mk_algebraMap R _ M ℤ _ _ m
+#align localization.mk_int_cast Localization.mk_int_cast
+
+end Localization
+
+namespace IsLocalization
+
+variable {K : Type _} [IsLocalization M S]
+
+theorem to_map_eq_zero_iff {x : R} (hM : M ≤ nonZeroDivisors R) : algebraMap R S x = 0 ↔ x = 0 := by
+  rw [← (algebraMap R S).map_zero]
+  constructor <;> intro h
+  · cases' (eq_iff_exists M S).mp h with c hc
+    rw [mul_zero, mul_comm] at hc
+    exact hM c.2 x hc
+  · rw [h]
+#align is_localization.to_map_eq_zero_iff IsLocalization.to_map_eq_zero_iff
+
+protected theorem injective (hM : M ≤ nonZeroDivisors R) : Injective (algebraMap R S) := by
+  rw [injective_iff_map_eq_zero (algebraMap R S)]
+  intro a ha
+  rwa [to_map_eq_zero_iff S hM] at ha
+#align is_localization.injective IsLocalization.injective
+
+protected theorem to_map_ne_zero_of_mem_nonZeroDivisors [Nontrivial R] (hM : M ≤ nonZeroDivisors R)
+    {x : R} (hx : x ∈ nonZeroDivisors R) : algebraMap R S x ≠ 0 :=
+  show (algebraMap R S).toMonoidWithZeroHom x ≠ 0 from
+    map_ne_zero_of_mem_nonZeroDivisors (algebraMap R S) (IsLocalization.injective S hM) hx
+#align is_localization.to_map_ne_zero_of_mem_non_zero_divisors IsLocalization.to_map_ne_zero_of_mem_nonZeroDivisors
+
+variable {S}
+
+theorem sec_snd_ne_zero [Nontrivial R] (hM : M ≤ nonZeroDivisors R) (x : S) :
+    ((sec M x).snd : R) ≠ 0 :=
+  nonZeroDivisors.coe_ne_zero ⟨(sec M x).snd.val, hM (sec M x).snd.property⟩
+#align is_localization.sec_snd_ne_zero IsLocalization.sec_snd_ne_zero
+
+theorem sec_fst_ne_zero [Nontrivial R] [NoZeroDivisors S] (hM : M ≤ nonZeroDivisors R) {x : S}
+    (hx : x ≠ 0) : (sec M x).fst ≠ 0 := by
+  have hsec := sec_spec M x
+  intro hfst
+  rw [hfst, map_zero, mul_eq_zero, _root_.map_eq_zero_iff] at hsec
+  · exact Or.elim hsec hx (sec_snd_ne_zero hM x)
+  · exact IsLocalization.injective S hM
+#align is_localization.sec_fst_ne_zero IsLocalization.sec_fst_ne_zero
+
+variable (S M) (Q : Type _) [CommRing Q] {g : R →+* P} [Algebra P Q]
+
+/-- Injectivity of a map descends to the map induced on localizations. -/
+theorem map_injective_of_injective (hg : Function.Injective g)
+    [i : IsLocalization (M.map g : Submonoid P) Q] :
+    --Porting note: Why does `i` need to be given explicitly?
+    Function.Injective (@map _ _ _ _ _ _ _ _ _ _ Q _ _ i g (Submonoid.le_comap_map M) : S → Q) := by
```

Comment:
Can you use the syntax `map (argname := i)` syntax instead of `@map _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ __ _ _ _ _ _ _ _ _ _i`?

Reply:
No because there is no argname for instances

Reply:
aah, good point